modules/edge-voting/src/voting.rs

```diff
@@ -0,0 +1,367 @@
+
+			let id = Self::vote_record_count() + 1;
+			<VoteRecords<T>>::insert(id, VoteRecord {
+				id: id,
+				is_commit_reveal: is_commit_reveal,
+				commitments: vec![],
+				reveals: vec![],
+				outcomes: outcomes,
+				winning_outcome: None,
+				tally: vec![],
+				data: VoteData {
+					initiator: _sender.clone(),
+					stage: VoteStage::PreVoting,
+					vote_type: vote_type,
+					tally_type: tally_type,
+				},
+			});
+
+			<VoteRecordCount<T>>::mutate(|i| *i += 1);
+			Self::deposit_event(RawEvent::VoteCreated(id, _sender, vote_type));
+			Ok(())
+		}
+
+		pub fn commit(origin, vote_id: u64, commit: [u8; 32]) -> Result {
+			let _sender = ensure_signed(origin)?;
+			let mut record = <VoteRecords<T>>::get(vote_id).ok_or("Vote record does not exist")?;
+			ensure!(record.is_commit_reveal, "Commitments are not configured for this vote");
+			ensure!(record.data.stage == VoteStage::Commit, "Vote is not in commit stage");
+			// TODO: Allow changing of commits before commit stage ends
+			ensure!(!record.commitments.iter().any(|c| &c.0 == &_sender), "Duplicate commits are not allowed");
+
+			// Add commitment to record
+			record.commitments.push((_sender.clone(), commit));
+			<VoteRecords<T>>::insert(record.id, record);
+			Ok(())
+		}
+
+		pub fn reveal(origin, vote_id: u64, vote: [u8; 32], secret: Option<[u8; 32]>) -> Result {
+			let _sender = ensure_signed(origin)?;
+			let mut record = <VoteRecords<T>>::get(vote_id).ok_or("Vote record does not exist")?;
+			ensure!(record.data.stage == VoteStage::Voting, "Vote is not in voting stage");
+			// Check vote is for a valid outcome
+			ensure!(record.outcomes.iter().any(|o| o == &vote), "Vote type must be binary");
+			// TODO: Allow changing of votes
+			ensure!(!record.reveals.iter().any(|c| &c.0 == &_sender), "Duplicate votes are not allowed");
+
+			// Ensure voter committed
+			if record.is_commit_reveal {
+				ensure!(record.commitments.iter().any(|c| &c.0 == &_sender), "Duplicate commits are not allowed");
+				let commit: (T::AccountId, [u8; 32]) = record.commitments
+					.iter()
+					.find(|c| &c.0 == &_sender)
+					.unwrap()
+					.clone();
+
+				let mut buf = Vec::new();
+				buf.extend_from_slice(&_sender.encode());
+				buf.extend_from_slice(&secret.unwrap().encode());
+				buf.extend_from_slice(&vote);
+				let hash = T::Hashing::hash_of(&buf);
+				ensure!(hash.encode() == commit.1.encode(), "Commitments do not match");
+			}
+
+			record.reveals.push((_sender.clone(), vote));
+			<VoteRecords<T>>::insert(record.id, record);
+			Ok(())
+		}
+
+		pub fn advance_stage_as_initiator(origin, vote_id: u64) -> Result {
+			let _sender = ensure_signed(origin)?;
+			let record = <VoteRecords<T>>::get(vote_id).ok_or("Vote record does not exist")?;
+			ensure!(record.data.initiator == _sender.clone(), "Invalid advance attempt by non-owner");
+			return Self::advance_stage(vote_id);
+		}
+
+		pub fn tally_as_initiator(origin, vote_id: u64) -> Result {
```

Comment:
Why not tally automatically when we shift to Completed?

Reply:
We can add it, I'm going to leave it out for now since I'm not fully certain where it should go. This is the most reasonable place.